x/evm/keeper/vote_handler.go

```diff
@@ -177,24 +180,29 @@ func (v voteHandler) HandleResult(ctx sdk.Context, result codec.ProtoMarshaler)
 	}
 
 	for _, event := range voteEvents.Events {
-		if err := handleEvent(ctx, ck, event, chain); err != nil {
+		if err := v.handleEvent(ctx, ck, event, chain); err != nil {
 			return err
 		}
 	}
 
 	return nil
 }
 
-func handleEvent(ctx sdk.Context, ck types.ChainKeeper, event types.Event, chain nexus.Chain) error {
-	// check if event confirmed before
-	eventID := event.GetID()
-	if _, ok := ck.GetEvent(ctx, eventID); ok {
-		return fmt.Errorf("event %s is already confirmed", eventID)
-	}
+func (v voteHandler) handleEvent(ctx sdk.Context, ck types.ChainKeeper, event types.Event, chain nexus.Chain) error {
 	if err := ck.SetConfirmedEvent(ctx, event); err != nil {
-		panic(err)
+		return err
 	}
-	ck.Logger(ctx).Info(fmt.Sprintf("confirmed %s event %s in transaction %s", chain.Name, eventID, event.TxID.Hex()))
+
+	switch event.GetEvent().(type) {
+	case *types.Event_ContractCall:
+		if err := v.handleContractCall(ctx, event); err != nil {
+			return err
+		}
+	default:
+		funcs.MustNoErr(ck.EnqueueConfirmedEvent(ctx, event.GetID()))
+	}
```

Comment:
I find it a bit weird to have this split where the Event_ContractCall is checked here, but the rest of the types is checked in the chain keeper. Can we do all checks in a single place?

Reply:
The type check in the chain keeper is more like a safeguard though, i.e. only allow supported types to go in. Here is the actual routing based on the type of the event.